src/middlewared/middlewared/plugins/pool_/dataset_encryption_info.py

```diff
@@ -356,17 +364,28 @@ async def export_keys_for_replication_internal(self, replication_task_id):
         source_mapping = await self.middleware.call(
             'zettarepl.get_source_target_datasets_mapping', task['source_datasets'], target_ds
         )
+        if include_encryption_root_children:
+            dataset_mapping = dataset_encryption_root_mapping or await self.dataset_encryption_root_mapping()
+        else:
+            dataset_mapping = {}
+
         for source_ds in task['source_datasets']:
             for ds_name, key in mapping[source_ds].items():
-                for dataset in await self.middleware.call(
-                    'pool.dataset.query', [['encryption_root', '=', ds_name]], {
-                        'extra': {'properties': ['encryptionroot']}
-                    }
-                ) if include_encryption_root_children else [{'id': ds_name}]:
+                for dataset in (dataset_mapping[ds_name] if include_encryption_root_children else [{'id': ds_name}]):
                     result[dataset['id'].replace(source_ds, source_mapping[source_ds], 1)] = key
 
         return result
 
+    @private
+    async def dataset_encryption_root_mapping(self):
+        dataset_encryption_root_mapping = collections.defaultdict(list)
+        for dataset in await self.middleware.call(
+            'pool.dataset.query', [], {'extra': {'properties': ['encryptionroot']}}
```

Comment:
So we now will query every dataset on the system if this arg is given to replication.query?

Reply:
That unfortunately needs to be done - on the positive side we are not getting all the properties so it does not has that much of an impact i.e my VM has 114 datasets and `replication.query` is taking now 0.16 seconds approximately.

Reply:
It was `0.09` before this change